### PR TITLE
Create ELB for master

### LINKED ISSUE #69: Remove Elastic IP, use ELB
As discussed, using an Elastic IP to assign a static, public IP address to the master(s) is a temporary solution, with the main drawback being that there's [a limit](https://docs.aws.amazon.com/general/latest/gr/aws_service_limits.html) on how many EIPs can be generated in a VPC.

The better solution is to setup a load balancer.

### PR DESCRIPTION
**Pull request by asymmetric**
Create an ELB in front of the master(s).

Also, not sure whether we should re-use LBs if they're already there, or if we should consider it an inconsistency and fail. I'm leaning more towards the second, since the LB that's already there has the same name but might have a completely different configuration.

Closes #69 .

### CODE REVIEW

File: resources/aws/elb.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		return false, microerror.MaskAny(err)
+	}
+
+	return true, nil
+}
+
+func (lb *ELB) CreateOrFail() error {
+	if lb.Client == nil {
+		return microerror.MaskAny(clientNotInitializedError)
+	}
```
Comment (JosephSalisbury) [inline resources/aws/elb.go]: Really specific nitpick, a new line here to match the above function would be cool

File: resources/aws/elb.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		return microerror.MaskAny(err)
+	}
+
+	return nil
+}
+
+func (lb *ELB) Delete() error {
+	if lb.Client == nil {
+		return microerror.MaskAny(clientNotInitializedError)
+	}
```
Comment (JosephSalisbury) [inline resources/aws/elb.go]: And a new line here

File: resources/aws/elb.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+func (lb *ELB) CreateOrFail() error {
+	if lb.Client == nil {
+		return microerror.MaskAny(clientNotInitializedError)
+	}
+	if _, err := lb.Client.CreateLoadBalancer(&elb.CreateLoadBalancerInput{
+		LoadBalancerName: aws.String(lb.Name),
+		Listeners: []*elb.Listener{
+			{
+				InstancePort:     aws.Int64(8080),
```
Comment (JosephSalisbury) [inline resources/aws/elb.go]: Where does this port come from?

**Comment by JosephSalisbury** [comment]
When would we have a situation where an ELB is already present?

File: resources/aws/elb.go (hunk 3 above)
Comment (asymmetric) [inline resources/aws/elb.go]: Hard-coded API server insecure port. Should be taken from the TPR, will change.

**Comment by asymmetric** [comment]
In case we fail to delete it for some reason:

* the [`DeleteFunc`](https://github.com/giantswarm/aws-operator/blob/master/service/create/service.go#L343) terminated before deleting the LB
* the operator crashed for whatever reason while/before deleting
* the API call to delete failed

I would argue these are error conditions and we should stop the creation of the cluster instead of reusing the LB.

**Comment by JosephSalisbury** [comment]
I think for now, stopping is fine. 

I'm not certain about how we'd present this to an human (that is, this cluster won't ever come up because of this ELB, please fix), but we can improve on this in the future. Alternatively, we do some work on ensuring resources are removed (either battle-testing, or some cleanup operator).

**Comment by asymmetric** [comment]
PTALA. :wave:

File: resources/aws/elb.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		return microerror.MaskAny(clientNotInitializedError)
+	}
+
+	if _, err := lb.Client.CreateLoadBalancer(&elb.CreateLoadBalancerInput{
+		LoadBalancerName: aws.String(lb.Name),
+		Listeners: []*elb.Listener{
+			{
+				InstancePort:     aws.Int64(int64(lb.InstancePort)),
+				LoadBalancerPort: aws.Int64(int64(lb.LoadBalancerPort)),
+				// TCP because we want to do SSL offloading, not termination
```
Comment (JosephSalisbury) [inline resources/aws/elb.go]: This is ssl passthrough, not offloading, right? I get confused by the terminology here.

Comment (asymmetric) [inline resources/aws/elb.go]: According to [AWS' terminology](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-listener-config.html), this is offloading:

> If you don't want the load balancer to handle the SSL termination (known as SSL offloading), you can use TCP for both the front-end and back-end connections, and deploy certificates on the registered instances handling requests.

Comment (asymmetric) [inline resources/aws/elb.go]: But yeah, if passthrough is the more standard term, let's use that.

Comment (asymmetric) [inline resources/aws/elb.go]: Or maybe I just read the quote wrong. ¯\_(ツ)_/¯.

**Comment by asymmetric** [comment]
PTALA.

File: resources/aws/elb.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+		return microerror.MaskAny(err)
+	}
+
+	// we have to populate some additional fields
+	lbDescription, err := lb.findExisting()
+	if err != nil {
+		return err
+	}
+
+	lb.setDNSFields(*lbDescription)
```
Comment (JosephSalisbury) [inline resources/aws/elb.go]: This can panic if `lbDescription` is `nil`, no?

File: resources/aws/kms.go [hunk 6]
```diff
@@ -17,6 +17,7 @@ func (kk *KMSKey) CreateIfNotExists() (bool, error) {
 }
 
 func (kk *KMSKey) CreateOrFail() error {
+	// TODO we should give it a name
```
Comment (JosephSalisbury) [inline resources/aws/kms.go]: Do we have an issue for tracking this?

File: resources/aws/recordset.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+func (record RecordSet) perform(action string) error {
+	if record.Client == nil {
+		return clientNotInitializedError
+	}
+
+	params := record.buildParams(action)
+
+	_, err := record.Client.ChangeResourceRecordSets(params)
+
```
Comment (JosephSalisbury) [inline resources/aws/recordset.go]: Please remove this newline #uselessnitpick

File: service/create/loadbalancer.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	if err := recordSet.Delete(); err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	s.logger.Log("debug", fmt.Sprintf("deleted DNS entry for ELB"))
+
+	// Delete HZ
+	if err := hz.Delete(); err != nil {
```
Comment (JosephSalisbury) [inline service/create/loadbalancer.go]: This feels weird to me. Are we going to have to redelegate the HZ whenever we create a cluster? I would expect the HZ to be kept around.

File: resources/aws/kms.go (hunk 6 above)
Comment (rossf7) [inline resources/aws/kms.go]: +1 to an issue for this and being pedantic they call it an alias in KMS.

File: service/create/loadbalancer.go (hunk 8 above)
Comment (asymmetric) [inline service/create/loadbalancer.go]: Good catch. I was thinking about this after work and planning on taking care of it in the morning ;)

File: resources/aws/elb.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	InstancePort     int
+}
+
+func (lb *ELB) CreateIfNotExists() (bool, error) {
+	if lb.Client == nil {
+		return false, microerror.MaskAny(clientNotInitializedError)
+	}
+
+	if err := lb.CreateOrFail(); err != nil {
+		if err.Error() == awsclient.ELBAlreadyExists {
```
Comment (nhlfr) [inline resources/aws/elb.go]: You will need to use `strings.Contains` here - this error was masked before, so its message will be not exactly equal to the message defined in the global variable.

File: resources/aws/recordset.go (hunk 7 above)
Comment (nhlfr) [inline resources/aws/recordset.go]: And bonus points from me for doing

```
if _, err := record.Client.ChangeResourceRecordSets(params); err != nil {
    return microerror.MaskAny(err)
}
```

Not a big deal, but looks cleaner IMO.

File: resources/aws/recordset.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	if err != nil {
+		return err
```
Comment (nhlfr) [inline resources/aws/recordset.go]: Ot at least mask this error. ;)

File: resources/aws/kms.go (hunk 6 above)
Comment (nhlfr) [inline resources/aws/kms.go]: https://github.com/giantswarm/aws-operator/issues/106

**Comment by asymmetric** [comment]
PTALA.

File: client/aws/errors.go [hunk 11]
```diff
@@ -3,12 +3,14 @@ package aws
...
-	KeyPairDuplicate        = "InvalidKeyPair.Duplicate"
-	SecurityGroupDuplicate  = "InvalidGroup.Duplicate"
+	AlreadyAssociated        = "Resource.AlreadyAssociated"
+	InvalidSubnetConflict    = "InvalidSubnet.Conflict"
+	BucketAlreadyExists      = "BucketAlreadyExists"
+	BucketAlreadyOwnedByYou  = "BucketAlreadyOwnedByYou"
+	KeyPairDuplicate         = "InvalidKeyPair.Duplicate"
+	SecurityGroupDuplicate   = "InvalidGroup.Duplicate"
+	ELBAlreadyExists         = "DuplicateLoadBalancerName"
+	ELBConfigurationMismatch = "already exists and it is configured with different parameters"
```
Comment (JosephSalisbury) [inline client/aws/errors.go]: Not a blocker, but the lack of uniformity here hurts me :D

Comment (nhlfr) [inline client/aws/errors.go]: This is the lack of uniformity of what AWS API returns, we don't have any control about it. Sad but true.

Comment (JosephSalisbury) [inline client/aws/errors.go]: Oh yeah, these strings are returned from aws. Brain fart. All good <3

**Pull request merged**